codegen/smithy-aws-go-codegen/src/main/java/software/amazon/smithy/aws/go/codegen/RestXmlProtocolGenerator.java

```diff
@@ -51,59 +66,238 @@ protected void writeMiddlewareDocumentSerializerDelegator(
+                });
+                writer.insertTrailingNewline();
+
+                if (targetShape.hasTrait(StreamingTrait.class)) {
+                    writer.write("v.$L = body", memberName);
+                    writer.write("return nil");
+                    return;
+                }
 
+                writer.addUseImports(SmithyGoDependency.IOUTIL);
+                writer.write("bs, err := ioutil.ReadAll(body)");
+                writer.write("if err != nil { return err }");
+                writer.openBlock("if len(bs) > 0 {", "}", () -> {
+                    if (targetShape.isBlobShape()) {
+                        writer.write("v.$L = bs", memberName);
+                    } else { // string
+                        writer.addUseImports(SmithyGoDependency.SMITHY_PTR);
+                        if (targetShape.hasTrait(EnumTrait.class)) {
+                            writer.write("v.$L = string(bs)", memberName);
+                        } else {
+                            writer.write("v.$L = ptr.String(string(bs))", memberName);
+                        }
+                    }
+                });
+                writer.write("return nil");
+            });
+        }
     }
 
-    @Override
-    public ShapeId getProtocol() {
-        return null;
+    // Writes middleware that delegates to deserializers for shapes that have explicit payload.
+    private void writeMiddlewarePayloadBindingDeserializerDelegator(GoWriter writer, Shape shape) {
+        String deserFuncName = ProtocolGenerator.getDocumentDeserializerFunctionName(shape, getProtocolName());
+        writer.write("err = $L(output, response.Body)", deserFuncName);
+        writer.openBlock("if err != nil {", "}", () -> {
+            writer.addUseImports(SmithyGoDependency.SMITHY);
+            writer.write(String.format("return out, metadata, &smithy.DeserializationError{Err:%s}",
+                    "fmt.Errorf(\"failed to deserialize response payload, %w\", err)"));
+        });
+    }
+
+    // Write middleware that delegates to deserializers for shapes that have implicit payload
+    private void writeMiddlewareDocumentBindingDeserializerDelegator(
+            GoWriter writer,
+            Shape shape,
+            String operand
+    ) {
+        writer.addUseImports(SmithyGoDependency.SMITHY_XML);
+        writer.addUseImports(SmithyGoDependency.SMITHY_IO);
+
+        writer.write("var decoder *smithyxml.NodeDecoder");
+        writer.write("buff := make([]byte, 1024)");
+        writer.write("ringBuffer := smithyio.NewRingBuffer(buff)");
+        writer.insertTrailingNewline();
+
+        writer.addUseImports(SmithyGoDependency.IO);
+        writer.write("body := io.TeeReader(response.Body, ringBuffer)");
+        writer.write("defer response.Body.Close()");
+        writer.insertTrailingNewline();
+
+        writer.openBlock("if decoder == nil {", "}", () -> {
```

Comment:
I think the `decoder` here might be from an older implementation, since the decoder here in this flow can